### PR TITLE
feat(resource): add resource limits cases

### PR DESCRIPTION
**Pull request by jijiechen**
1. Run the control plane and connect specified numbers of DPs onto it and assert it should not fail.
2. Compare the runnings between settings of `GOMEMLIMIT`.

```sh
# set the amount of memory limit onto the CP workload in MiB
export LIMIT_MEGA_MEMORY=256
# set the amount of CPU limit onto the CP workload in millicores
export LIMIT_MILLI_CPU=1000

# set the numbers of services
export PERF_TEST_NUM_SERVICES=5
# set the numbers of instances per service
export PERF_TEST_INSTANCES_PER_SERVICE=20

make run
```

Cases included:
1. should deploy all services and instances
2. should crash when deploy all services and instances with half CP resource
3. should not crash when control plane has `GOMEMLIMIT` set when all services and instances with half CP resource

A lot of metrics are periodically fetched and printed to output, so I'm trying to save it to a file, otherwise it'll be troublesome to open the view for the GHA runs.

### CODE REVIEW

File: test/k8s/res_limits_test.go [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/ghodss/yaml"
+	"github.com/gruntwork-io/terratest/modules/logger"
+	"k8s.io/apimachinery/pkg/watch"
+	"os"
```
Comment (Icarus9913) [inline test/k8s/res_limits_test.go:11]: format

**Comment by Icarus9913** [comment]
Almost LGTM

File: test/k8s/res_limits_test.go [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+						op = "add"
+						idxStr = "-"
+					}
+
+					patchJson += fmt.Sprintf(`,{"op": "%s", "path": "/spec/template/spec/containers/0/env/%s", "value": {"name": "GOMEMLIMIT", "value":"%dMiB"}}`,
+						op, idxStr, memLimit)
+				}
+			}
+
+			patchJson += "]"
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:169]: This is a bit hard to follow. How about we change `patchJson` to `[]string`
Then do
```
patchJson := []string{
  fmt.Sprintf("{"op": "replace", "path": "/spec/template/spec/containers/0/resources/requests/cpu", "value": "%dm"}, miliCPU)
...
}
```

then do
```
"--patch", fmt.Sprintf("[%s]", strings.Join(patchJson, ","))
```

File: test/k8s/res_limits_test.go [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+							break
+						}
+					}
+
+					op := "replace"
+					idxStr := fmt.Sprintf("%d", idx)
+					if idx == -1 {
+						op = "add"
+						idxStr = "-"
+					}
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:162]: Do we need to handle -1 idx and `-` even? Can we simplify the code by assuming that it was added already because `goMemLimitAdded` tracks this

File: test/k8s/res_limits_test.go [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/limits/memory", "value": "%dMi"}`
+
+			if !goDebugAdded {
+				goDebugAdded = true
+				patchJson += `,{"op": "add", "path": "/spec/template/spec/containers/0/env/-", "value": {"name": "GODEBUG", "value":"gctrace=1"}}`
+			}
+
+			if addGoMemLimitEnv {
+				memLimit := memMega - 50
+				if memLimit < 10 {
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:131]: Can it be lower than 10? Why 10 is minimum?

File: test/k8s/res_limits_test.go [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+				for _, cts := range pod.Status.ContainerStatuses {
+					_, _ = fmt.Fprintf(GinkgoWriter, "Pod %s - %s: %s\n", pod.Name, cts.Name, cts.State.String())
+				}
+			}
+		})
+	})
+
+}
+
+func printLog(msg string, args ...interface{}) {
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:500]: Just wondering. Was `framework.Logf` not enough?

File: test/k8s/res_limits_test.go [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+					return
+				}
+
+				if created {
+					printLog("%d pods are now all created\n", expectedNumOfPods)
+				} else {
+					printLog("only %d of %d pods created\n", len(pods), expectedNumOfPods)
+					x <- created
+					return
+				}
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:235]: wouldn't it be easier to pass error channel here and do
```
if !created {
  errCh <- errors.Errorf("only %d of %d pods created", len(pods), expectedNumOfPods)
  return
}
Logf("%d pods are now all created\n", expectedNumOfPods)
```


at the end of the func just pass errCh<-nil?

File: test/k8s/res_limits_test.go [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+				wg.Wait()
+				x <- true
+			}(waitCh)
+
+			go func() {
+				val := <-waitCh
+				if val {
+					printLog("data plane pods are now all available, entering the stabilization sleep")
+					time.Sleep(stabilizationSleep)
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:264]: can't we just sleep before `x <- true`?

File: test/k8s/res_limits_test.go [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+						args := []string{"logs", "-c", "control-plane", pod.Name}
+						if hasPodContainerCrashed(pod) {
+							args = append(args, "-p")
+						}
+
+						podLogs, err := k8s.RunKubectlAndGetOutputE(
+							cluster.GetTesting(),
+							cluster.GetKubectlOptions(Config.KumaNamespace),
+							args...)
+						Expect(err).ToNot(HaveOccurred())
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:335]: shouldn't `Cluster.GetKumaCPLogs` just work?

File: test/k8s/res_limits_test.go [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+
+		// should not crash, dump memory usage
+		It("should deploy all services and instances", func() {
+
+			adjustResource(cpu, memory, false)
+
+			deployDPs()
+
+			errCh := make(chan error)
+			ctx, cancelMonitoring := context.WithCancel(context.Background())
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:392]: We usually do `defer cancelMonitoring()` so we don't forget to call it on unexpected exists. For example what if `Expect(dpRet).To(BeTrue(), "data planes should be run and available")` fails?

File: test/k8s/res_limits_test.go [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+			case <-dpCh:
+				printLog("dpCh returned\n")
+				break
+			case err = <-errCh:
+				printLog("errCh returned\n")
+				break
+			}
+			cancelMonitoring()
+			Expect(err).To(HaveOccurred(), "control plane should crash with half resource")
+			if err != nil {
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:436]: no need to if it. If it's nil, then `Expect(err).To(HaveOccurred` will fail

File: test/k8s/res_limits_test.go [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+				_, _ = fmt.Fprint(GinkgoWriter, err.Error())
+			}
+
+			cpPods, err := k8s.ListPodsE(cluster.GetTesting(), cluster.GetKubectlOptions(Config.KumaNamespace), metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("app=%s", Config.KumaServiceName),
+			})
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:442]: how about `cluster.GetKuma().GetKumaCPPods`

File: test/k8s/res_limits_test.go [hunk 12]
```diff
@@ -0,0 +1,512 @@
...
+
+			dpCh := make(chan bool)
+			waitForDPs(dpCh)
+
+			var err error
+			select {
+			case dpRet := <-dpCh:
+				printLog("dpCh returned\n")
+				Expect(dpRet).To(BeTrue(), "data planes should be run and available")
+				break
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:403]: do we need break? same for other selects

**Review by jakubdyszkiewicz — CHANGES_REQUESTED** [review]

File: test/k8s/res_limits_test.go (hunk 1 above)
Comment (jijiechen) [reply test/k8s/res_limits_test.go:11]: `make check` will validate the format. So I assume it's good.

File: test/k8s/res_limits_test.go [hunk 13]
```diff
@@ -0,0 +1,512 @@
...
+			Setup(cluster)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	BeforeEach(func() {
+		Expect(framework.ReportSpecStart(cluster)).To(Succeed())
+	})
+
+	E2EAfterAll(func() {
+		Expect(cluster.TriggerDeleteNamespace(TestNamespace)).To(Succeed())
```
[the remark this answers is not in the document]
Comment (jijiechen) [reply test/k8s/res_limits_test.go:71]: Right, it's already deleted in `AfterEach`

File: test/k8s/res_limits_test.go [hunk 14]
```diff
@@ -0,0 +1,512 @@
...
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/limits/cpu", "value": "%dm"},
+{"op": "replace", "path": "/spec/template/spec/containers/0/resources/limits/memory", "value": "%dMi"}`
+
+			if !goDebugAdded {
+				goDebugAdded = true
+				patchJson += `,{"op": "add", "path": "/spec/template/spec/containers/0/env/-", "value": {"name": "GODEBUG", "value":"gctrace=1"}}`
+			}
+
+			if addGoMemLimitEnv {
+				memLimit := memMega - 50
```
[the remark this answers is not in the document]
Comment (jijiechen) [reply test/k8s/res_limits_test.go:130]: It's better to set `GOMEMLIMIT` as a value slightly lower than the container's memory. 
`GOMEMLIMIT` is a limitation that the Go runtime will set on the consumption of the process. The container hosting the process itself also needs to consume some resource. To make sure the self-limitation mechanism senstive we want to make sure the `GOMEMLIMIT` is a little lower than all available memory of the whole container.
`50Mi` is only a relatively small number picked randomly. A better solution is to use a percentage, like `10%`.

File: test/k8s/res_limits_test.go (hunk 2 above)
Comment (jijiechen) [reply test/k8s/res_limits_test.go:169]: Great improvement!

File: test/k8s/res_limits_test.go (hunk 8 above)
Comment (jijiechen) [reply test/k8s/res_limits_test.go:335]: I didn't use `GetKumaCPLogs` before, thanks for that note. Just had a check, I think it's slightly different here. 
The logic here is monitoring every single pod and dump it's log only if it fails and `GetKumaCPLogs` will dump logs from all CP pods.

Comment (jijiechen) [reply test/k8s/res_limits_test.go:335]: I'll mark this thread as resolved FTM, feel free to open it again.

File: test/k8s/res_limits_test.go (hunk 9 above)
Comment (jijiechen) [reply test/k8s/res_limits_test.go:392]: Sure, from coding perspective, it's a good practice to calling it using `defer`.

File: test/k8s/res_limits_test.go (hunk 11 above)
Comment (jijiechen) [reply test/k8s/res_limits_test.go:442]: Here is just to print container statuses of failed pods. So not logs. Logs are captured and printed in the watching.

File: test/k8s/res_limits_test.go (hunk 4 above)
Comment (jijiechen) [reply test/k8s/res_limits_test.go:131]: It's a basic protection to the value making sure it's not too low. I'll use 90% instead.

File: test/k8s/res_limits_test.go (hunk 3 above)
Comment (jijiechen) [reply test/k8s/res_limits_test.go:162]: `goMemLimitAdded` means whether we want to add the env var in this call, while this `idx` tracks if the env var exists in the deployment.

File: test/k8s/res_limits_test.go (hunk 5 above)
Comment (jijiechen) [reply test/k8s/res_limits_test.go:500]: Oh, I didn't know this helper function. Let me use it.

File: test/k8s/res_limits_test.go (hunk 6 above)
Comment (jijiechen) [reply test/k8s/res_limits_test.go:235]: Yeah, it can be much simplified in this way.
Thanks.

File: test/k8s/res_limits_test.go [hunk 15]
```diff
@@ -0,0 +1,487 @@
...
+			err := k8s.RunKubectlE(
+				cluster.GetTesting(),
+				cluster.GetKubectlOptions(Config.KumaNamespace),
+				"patch", "deployment", Config.KumaServiceName,
+				"--type=json",
+				"--patch", fmt.Sprintf("[%s]", strings.Join(patchJson, ",")),
+			)
+			Expect(err).ToNot(HaveOccurred())
+
+			time.Sleep(2 * time.Second)
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:175]: Why do we sleep? Can status deployment not be available? If so, wouldn't we need Eventually?

File: test/k8s/res_limits_test.go [hunk 16]
```diff
@@ -0,0 +1,485 @@
...
+			var err error
+			select {
+			case <-dpCh:
+				Logf("dpCh returned\n")
+			case err = <-errCh:
+				Logf("errCh returned\n")
+			}
+			cancelMonitoring()
+			Expect(err).ToNot(HaveOccurred(), "control plane should not crash")
+			if err != nil {
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:456]: There is no way we'll get there. Ginkgo will stop execution of this `It(` when err is not nil on this line `Expect(err).ToNot(HaveOccurred(), "control plane should not crash")`

File: test/k8s/res_limits_test.go [hunk 17]
```diff
@@ -0,0 +1,485 @@
...
+			cpPods, err := k8s.ListPodsE(cluster.GetTesting(), cluster.GetKubectlOptions(Config.KumaNamespace), metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("app=%s", Config.KumaServiceName),
+			})
+			Expect(err).ToNot(HaveOccurred())
+			for _, pod := range cpPods {
+				if k8s.IsPodAvailable(&pod) {
+					continue
+				}
+
+				_, _ = fmt.Fprintf(GinkgoWriter, "Pod %s: %s\n", pod.Name, pod.Status.Phase)
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:428]: Why are we printing this? I think we won't even get there? We are printing pods of Kuma CP. If it was not available we would stop execution at
```
Expect(err).To(HaveOccurred(), "control plane should crash with half resource")
```
no?

File: test/k8s/res_limits_test.go [hunk 18]
```diff
@@ -0,0 +1,485 @@
...
+			Expect(err).ToNot(HaveOccurred())
+			for _, pod := range dpPods {
+				if k8s.IsPodAvailable(&pod) {
+					continue
+				}
+
+				_, _ = fmt.Fprintf(GinkgoWriter, "Pod %s: %s\n", pod.Name, pod.Status.Phase)
+				for _, cts := range pod.Status.ContainerStatuses {
+					_, _ = fmt.Fprintf(GinkgoWriter, "Pod %s - %s: %s\n", pod.Name, cts.Name, cts.State.String())
+				}
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:470]: Why are we printing this? Should this be assertion or what's the reason?

File: test/k8s/res_limits_test.go [hunk 19]
```diff
@@ -0,0 +1,485 @@
...
+
+			dpCh := make(chan bool)
+			waitForDPs(dpCh)
+
+			var err error
+			select {
+			case <-dpCh:
+				Logf("dpCh returned\n")
+			case err = <-errCh:
+				Logf("errCh returned\n")
```
Comment (jakubdyszkiewicz) [inline test/k8s/res_limits_test.go:452]: Is just printing enough or should this be an assertion?

Comment (jijiechen) [reply test/k8s/res_limits_test.go:452]: There is an assertion underneath.

File: test/k8s/res_limits_test.go (hunk 18 above)
Comment (jijiechen) [reply test/k8s/res_limits_test.go:470]: I want to print dp statuses to help diagnose issues. For the case when "control plane should not crash", the DPs are also expected to be healthy. So if there is any DP unready, it should be not related to the CP.

File: run-limit.log
Comment (lahabana) [inline run-limit.log:1]: You checked this in by accident

**Comment by jijiechen** [comment]
The code changes are now complete and could you take a second review? @jakubdyszkiewicz

File: test/k8s/suite_test.go [hunk 20]
```diff
@@ -103,5 +103,6 @@ var _ = ReportAfterSuite("compile report", func(ginkgoReport Report) {
 })
 
 var (
-	_ = Describe("Simple", Simple, Ordered)
+	_ = XDescribe("Simple", Simple, Ordered)
```
Comment (Automaat) [inline test/k8s/suite_test.go:106]: why are we disabling simple test case?

**Review by Automaat — APPROVED** [review]

File: test/framework/silent_kubectl/client.go [hunk 21]
```diff
@@ -0,0 +1,44 @@
+// nolint:all  // code here is extracted and revised from github.com/gruntwork-io/terratest
```
Comment (lobkovilya) [inline test/framework/silent_kubectl/client.go:1]: why do you extracting the code instead of just adding the `github.com/gruntwork-io/terratest` as dependency?

**Review by jakubdyszkiewicz — COMMENTED** [review]
the test is ok, but I have the same questions as Ilya and Marcin

File: test/k8s/suite_test.go (hunk 20 above)
Comment (jijiechen) [reply test/k8s/suite_test.go:106]: This is a mistake. I forgot why I did this. Probably I was trying to focus on my own case (well, I can actually use `F`)

File: test/framework/silent_kubectl/client.go (hunk 21 above)
Comment (jijiechen) [reply test/framework/silent_kubectl/client.go:1]: I duplicated these functions to remove the redundant and verbose logging. ("Configuring kube client using..." blabla, this logging is polluting the test output very much since I need to poll the cluster)
I'll remove these duplications once `terratest` supports configure these verbosity. (Charly has a PR to do it, I'll find some time and help it merged)

Comment (jijiechen) [reply test/framework/silent_kubectl/client.go:1]: And that's why I call it "silent kubectl", because the one from the module talks too much.

**Comment by jijiechen** [comment]
I'll merge it since we've aligned with the test and I've resolved all the conversations.
